### PR TITLE
enhancement(journald source): Allow remapping priority numbers to words

### PR DESCRIPTION
**Pull request by bruceg**
Signed-off-by: Bruce Guenter <bruce@timber.io>

[Reference](https://github.com/timberio/vector/issues/2678#issuecomment-639871186) #2678

### CODE REVIEW

**Comment by bruceg** [comment]
I'd like opinions on whether to use abbreviated or full terms, and lower or upper case. I could make the `remap_priority` an enum to choose, but it seems overkill.

**Review by ktff — APPROVED** [review]
LGTM

Enum does seem like an overkill. 

Abbreviated lowercase according to [this page](https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html), although I'm not sure how it's in practice.

**Comment by bruceg** [comment]
> Abbreviated lowercase according to [this page](https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html), although I'm not sure how it's in practice.

Yes, I saw that, and that was my question—is it seen in other places in practice and what standard exists?

**Comment by binarylogic** [comment]
Thanks, @bruce. I appreciate that this helps solve #2678, but I really dislike this approach as a whole. If we continue down this path, sources and sinks will be littered with all kinds of edge case schema options. You can already see this with this new option and all of the `*_key` options throughout Vector.

The proper solution lies in how we approach schemas in Vector (https://github.com/timberio/vector/pull/2428). For example, if we had event metadata that told us the `PRIORITY` field contains known priorities, then we could use that to map the severity in the `gcp_stackdriver_logs` sink. The `gcp_stackdriver_logs` sink could go down a list of rules to try and determine the severity. Does that make sense?

In other words, we could implement a function like:

```
func getPriorities() {
  // look for priorities field
  // if that doesn't exist, look for a severities field
  // if that doesn't exist, look for a level field
  // if that doesn't exist, return nil
}
```

A log event could use internal metadata to help find this information. This metadata would largely be set at the source and would not require changing the actual log data.

Just food for thought as we continue to encounter schema related issues like this.

**Comment by lukesteensen** [comment]
@binarylogic I don't think this solution and your theoretical schema-based solution are really all that different.

> If we continue down this path, sources and sinks will be littered with all kinds of edge case schema options. You can already see this with this new option and all of the `*_key` options throughout Vector.

The `_key` options are pretty different from what's happening here. They specify where to find data, which we'll always need some form of (i.e. a mapping from input data to internal representation). This option is more of a "how" related to storage.

> For example, if we had event metadata that told us the `PRIORITY` field contains known priorities, then we could use that to map the severity in the `gcp_stackdriver_logs` sink.

Unfortunately, we don't currently have event metadata so we can't do this :smile: 

With event metadata, I think we'd likely just write a normalized priority representation to that metadata rather than a pointer to something that'd then need to be interpreted. In that view, it's not all that different from what's happening here. We're just using the event data instead of metadata because that's what's available.

**Comment by binarylogic** [comment]
Those are all good points. To clarify, I'm not pushing for a solution in this PR, just trying to increase visibility on something that I think we should address in the near future. And also linking the issues together so we have a catalog when we want to approach it.

Like a wise man once said, "all good solutions start with a healthy amount of complaining".

**Review by binarylogic — APPROVED** [review]
